proposals/readonly-struct-init.md

```diff
@@ -0,0 +1,161 @@
+```
+
+This does mean that the feature will only apply to auto-implemented properties that are 
+re-compiled with a new compiler. The compiler today emits the fields as `private readonly` and 
+hence will be inaccessible to calling code. Upon recompile the backing fields will be emitted 
+as specified above.
+
+### Verification
+
+As is the case with several features added since C# 7.2 this feature will not pass any known 
+peverify implementation as it makes no real distinction between read only and mutable memory. 
+However it is still desirable for C#, and other languages, to emit IL that can be verified 
+with appropriate adjustments to the verification spec. Eventually allowing for projects like
+[ILVerify](https://github.com/dotnet/corert/tree/master/src/ILVerify) to take the place of 
+peverify today. 
+
+This rules could be adjusted to allow for mutation of a `readonly` field following an `.initobj` 
+or `.call .ctor` instruction if:
+
+1. The reciever of the call is a local 
+1. The type of the reciever is a `struct` or `ref struct`
+1. There are no `ref` locals which refer to that local
+
+The `readonly` fields would remain mutable until one of the following occured:
+
+1. The reciever was passed as any `ref` to another function except when being used as `this`
+1. A `ref` local was taken to the local or any of its members
+
+This may force conforming compilers to create extra locals in some cases to allow for a
+successful verification of the emitted code.
+
+## Considerations
+
+### Type Parameters
+
+The feature does not apply to type parameters constrained to a `struct`. There is just no case 
+where such a type would have a field (`readonly` or not) or an auto-implemented property. In the
+future if we allow for `struct` inheritance it's possible this could change and the decision 
+re-evaluated.
+
+## classes
+
+This can't apply to `class` instances as it's much easier to smuggle a reference out of the 
+constructor. The `this` value can be easily stored in a `static` making it accessible to the 
+calling code long before the object initialization has completed. 
+
+``` csharp
+static C Global;
+
+class C { 
+    public readonly int Field;
+
+    public C() { 
+        Global = this;
+    }
+}
+
+int SomeMethod() => Global.Field++;
+
+// The instance is observed by caller while mutations are still in progress.
+var c = new C() { Field = SomeMethod()}
+```
+
+A `struct` can't be smuggled out in such a way as it would require a direct copy to another 
+instance of the same type or by boxing to an interface / object. In either case it would be a 
+different value and not interesting here.
+
+### Future Considerations
+
+There is another proposal to allow for simple data types in the language via the `data` modifier.
+Such types would have auto-matic equality, hashing, ToString, etc ... Combined with this proposal
+and immutable data now has the simplest possible definition in C#:
```

Comment:
I don't think `auto-matic` is normally hyphenated.
`Combined with this proposal, immutable data` (no `and`)?